volatility3/framework/plugins/linux/bash_env.py

```diff
@@ -0,0 +1,227 @@
+            ):
+                if key != "" or val != "":
+                    varstr = varstr + key + "=" + val + " "
+            yield (0, (task.pid, utility.array_to_string(task.comm), varstr))
+
+    def bash_environment(self, task, pack_format, hash_json_file, proc_layer_name):
+        for (key, val) in self.dynamic_env(task, pack_format, proc_layer_name):
+            yield (key, val)
+
+        for (key, val) in self.shell_variables(
+            task, pack_format, hash_json_file, proc_layer_name
+        ):
+            yield (key, val)
+
+    def dynamic_env(self, task, pack_format, proc_layer_name):
+        proc_layer = self.context.layers[proc_layer_name]
+        addr_cache = {0: 1}
+        addr_sz = 8
+        if pack_format == "I":
+            addr_sz = 4
+        for vma in task.mm.get_mmap_iter():
+            if not (vma.vm_file and vma.get_protection() == "rw-"):
+                continue
+            fname = vma.get_name(self.context, task)
+            if fname.find("ld") == -1 and (
+                not fname.endswith(("/bin/bash", "/bin/dash", "/bin/sh"))
+            ):
+                continue
+            env_start = 0
+            vma_start = int(vma.vm_start)
+            vma_end = int(vma.vm_end)
+            vma_len = vma_end - vma_start
+            vma_data = proc_layer.read(vma_start, vma_len, pad=True)
+
+            for off in range(0, vma_len - addr_sz, addr_sz):
+                addrstr = vma_data[off : off + addr_sz]
+                addr = struct.unpack(pack_format, addrstr)[0]
+
+                if addr in addr_cache:
+                    continue
+                addr_cache[addr] = 1
+
+                if addr:
+                    firstaddrstr = proc_layer.read(addr, addr_sz, pad=True)
+                    if not firstaddrstr or len(firstaddrstr) != addr_sz:
+                        continue
+                    firstaddr = struct.unpack(pack_format, firstaddrstr)[0]
+                    buf = proc_layer.read(firstaddr, 64, pad=True)
+                    if not buf:
+                        continue
+                    eqidx = buf.find(b"=")
+                    if eqidx > 0:
+                        nullidx = buf.find(b"\x00")
+                        if nullidx >= eqidx:
+                            env_start = addr
+
+            if env_start == 0:
+                continue
+            vmlinux = self.context.modules[self.config["kernel"]]
+            envars = self.context.object(
+                vmlinux.symbol_table_name + constants.BANG + "array",
+                layer_name=proc_layer_name,
+                offset=env_start,
+                subtype=self.context.symbol_space.get_type(
+                    vmlinux.symbol_table_name + constants.BANG + "pointer"
+                ),
+                count=256,
+            )
+            for var in envars:
+                if var:
+                    sizes = [8, 16, 32, 64, 128, 256, 384, 512, 1024, 2048, 4096]
```

Comment:
I believe the check on line 140 "idx < eqidx:" handles the null before the =

Reply:
I agree the reading over and over based on the `sizes` list seems inefficient, but just reading the whole page wouldn't be a good fix as 1) `var` is likely not page aligned and 2) the environment can span page boundaries. The code attempts to find the smallest valid read that finds a complete variable key/data pair. It seems like 135 could be a `break` though because if reading a smaller size fails then larger reads will also fail. Thoughts?